device-identifier/shbdeviceidentifier/utilities/capture_utilities.py

```diff
@@ -150,3 +215,195 @@ def convert_Capture_to_Line(cap: Capture, measurement: str = "packet") -> List[s
+        args = shlex.split(capinfos_path.as_posix() + " " + file_path.as_posix())
+
+        # Run tshark and get output as text
+        res = subprocess.run(args, capture_output=True, text=True).stdout
+
+        # Parse output into dictionary
+        return _capinfos_string_to_dict(res)
+
+
+def _capinfos_string_to_dict(capinfos: str) -> Optional[Dict]:
+    capinfos = capinfos.split("\n")[:-1]
+    capinfos_dict = {}
+
+    for line in capinfos[:-7]:
+        try:
+            k, v = line.split(":", 1)
+            k = k.strip().replace(" ", "_")
+            capinfos_dict[k] = v.strip()
+        except ValueError as e:
+            logger.warning(f"Could not parse capinfos line: {line}")
+            logger.debug(e)
+
+    # TODO: add support for multiple interfaces
+    # Add interface info
+    interface_dict = {}
+    for line in capinfos[-6:]:
+        try:
+            k, v = line.split("=", 1)
+            k = k.strip().replace(" ", "_")
+            interface_dict[k] = v.strip()
+        except ValueError as e:
+            logger.warning(f"Could not parse capinfos interface line: {line}")
+            logger.debug(e)
+
+    capinfos_dict["interfaces"] = [interface_dict]
+
+    return capinfos_dict if capinfos_dict else None
+
+
+@nb.jit(forceobj=True)
+def _get_data_len_from_scapy_packet(packet: Packet) -> np.uint16:
+    """
+    Returns the length of the data in a scapy packet in bytes.
+    """
+    # TODO: decide which length to return
+    # Currently returning the size of the Ethernet Frame!
+    return np.uint16(len(packet))
+
+
+@nb.jit(forceobj=True)
+def _get_protocol_from_scapy_packet(packet: Packet) -> str:
+    """
+    Returns the transport layer protocol name of a scapy packet.
+    """
+    if TCP in packet:
+        return "tcp"
+    elif UDP in packet:
+        return "udp"
+    else:
+        return ""
+
+
+@nb.jit(forceobj=True)
+def _get_timestamp_from_scapy_packet(packet: Packet) -> pd.Timestamp:
+    """
+    Returns the timestamp of a packet in nanoseconds.
+    """
+    time = pd.NaT
+
+    if hasattr(packet, "time"):
+        # Cannot use fromtimestamp with microseconds, so we split the timestamp up and
+        # merge them again explicitly afterwards
+        timestamp, fractions = str(packet.time).split(".")
+        time = pd.Timestamp.fromtimestamp(int(timestamp))
+
+        time = pd.Timestamp(
+            year=time.year,
+            month=time.month,
+            day=time.day,
+            hour=time.hour,
+            minute=time.minute,
+            second=time.second,
+            microsecond=int(fractions),
+        )
+    return time
+
+
+@nb.jit(forceobj=True)
+def _get_address_from_scapy_packet(packet: Packet, src: bool = True) -> str:
+    """
+    Returns the source address and port of a scapy packet.
+    Per default the source address is returned. Set src to `False` to get the destination address.
+    """
+    addr: str = ""
+
+    if IP in packet:
```

Comment:
We have very deep nesting here. How about
```python
if not IP in packet: return addr
if src:
    ...
```

Reply:
fixed